### PR TITLE
Add support for SRv4 Power Board Serial Interface

### PR DESCRIPTION
**Pull request by raccube**
## Checklist

- [x] I have updated the relevant documentation
- [x] I have added the `semver-major`, `semver-minor` or `semver-patch` label as appropriate
- [x] I would like multiple reviewers to approve my code before merging

## Why do you want to make these changes?

The new SR Power Board firmware uses a new serial protocol for communication with the power board.

## Which parts of the codebase do your changes affect?

SR v4 Power Board. This is not yet fully integrated as we need to detect which `PowerBoard` class to use based on firmware version.

## How do your changes affect student or volunteer experience?

Once fully integrated, this will allow use of updated power boards.

## Are your changes related to any existing issues or PRs? How so?

Part of #809

### CODE REVIEW

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (8BitJosh) [inline j5/backends/hardware/sr/v4/serial/power_board.py:236]: This should divide the response by 1000. The serial returns the value in mA and this function says it returns Amps

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (8BitJosh) [inline j5/backends/hardware/sr/v4/serial/power_board.py:199]: There are now 7 outputs, the 7th (identifier/index 6) being the 5V output as this is no longer used by the brain. This output supports the same functionality as the others so this can just be extended.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (8BitJosh) [inline j5/backends/hardware/sr/v4/serial/power_board.py:316]: Voltage needs dividing by 1000. Board returns mV function returns Volts

Comment (raccube) [reply j5/backends/hardware/sr/v4/serial/power_board.py]: [c3989df](https://github.com/j5api/j5/pull/812/commits/c3989df945938fdcedc7e5ba6ff411edbaeb3df3)

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (raccube) [reply j5/backends/hardware/sr/v4/serial/power_board.py:148]: [c3989df](https://github.com/j5api/j5/pull/812/commits/c3989df945938fdcedc7e5ba6ff411edbaeb3df3)

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (raccube) [reply j5/backends/hardware/sr/v4/serial/power_board.py]: 827cc89

Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py]: It's really not that simple sadly. We need to update the board too to support this and turn on the 5V output.

Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py]: We can do this after #814, for now I'd suggest we don't support the 5V output in this PR.

Comment (raccube) [reply j5/backends/hardware/sr/v4/serial/power_board.py:111]: I dropped the commit that changed it to 7. We can deal with this separately.

Comment (WillB97) [reply j5/backends/hardware/sr/v4/serial/power_board.py:111]: Note that this will prevent the 5 volt output from being enabled.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/power_board.py:174]:
```suggestion
        duration_ms = round(duration.total_seconds() * 1000)
```

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/power_board.py:175]: Rember the sign bit
```suggestion
        max_duration = (2 ** 31) - 1  # int32 max
```

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/power_board.py:220]: Since the start button's state is latching you want to discard the first response value otherwise this returns immediately if the start button has been pressed at any point since it was last checked.

**Review by WillB97 — COMMENTED** [review]
Mostly there, still has a few bits to sort out.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (raccube) [reply j5/backends/hardware/sr/v4/serial/power_board.py:223]: Didn't initially understand what you meant by "latched" but that makes sense.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py:117]: The frontend components don't support the 5V output yet, so we will implement it later. Also blocked on #814

File: j5/backends/hardware/sr/v4/serial/protocol.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/protocol.py]: Fixed in 3bd3e1bfe6c66277e2b9ce63e15f3a648137fd6c

File: j5/backends/hardware/sr/v4/serial/protocol.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/protocol.py:78]: The first digit here is the hardware version, the second segment is the major version.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/power_board.py:182]: This seems horribly inefficient. The only improvement the original line needed was to check that duration wasn't negative.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/power_board.py:188]: Same as above.

**Review by WillB97 — COMMENTED** [review]
There seems to have been some strange changes made here.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py:182]: Sorry, I'm not sure what you mean by this being inefficient?

File: j5/backends/hardware/sr/v4/serial/protocol.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/protocol.py:78]: This would be useful to have documented in the protocol docs

Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/protocol.py:78]: 0a20492

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [reply j5/backends/hardware/sr/v4/serial/power_board.py:182]: At first glance here it appears that you are about to do 2 billion iterations to find out if this generator contains your value. It does appear that range impliments a contains method to shortcut this (still ~7x slower than the straight compare). However the exclusive top end of this range makes it arguably less readable than simply comparing it's bounds.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py:117]: We can do this once #817 is merged

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py:182]: Interesting that it's that much slower, usually using a range object is just as fast (There's a bytecode op for it)

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py:188]: 5f487ca

**Review by trickeydan — APPROVED** [review]
Approved 🎉

I'd like to wait until @WillB97 or @8BitJosh approves before merging this too.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/power_board.py:29]: Major version

File: j5/backends/hardware/sr/v4/serial/protocol.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/protocol.py:151]: This duplicates the functionality of the caller using query.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py]: 4793d30307f2b6a5dc67de765c8c9eae5443e128

File: j5/backends/hardware/sr/v4/serial/protocol.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/protocol.py]: d5572689db1922ca334f3b67c76d29f555a8a819

File: j5/backends/hardware/sr/v4/serial/protocol.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/protocol.py:106]: ACK is still a response, we can leave handling responses to the caller.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/power_board.py:188]: This range excludes 0 which would be sent to silence the buzzer.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/power_board.py:102]: Is this actually clearer than just explicitly creating the dict fo 0 & 1?

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/power_board.py:139]: Should this now be 7?

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/power_board.py:153]: Should this now be 7?

File: j5/backends/hardware/sr/v4/serial/protocol.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/protocol.py:151]: So that we don't have to check if they are valid integers later. e.g `hh.ty` would not be caught until a TypeError occurs later in the program.

File: j5/backends/hardware/sr/v4/serial/protocol.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/protocol.py:106]: Given that the caller in this instance is still within the class, I would rather handle this case once rather than many times.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py:188]: 0 isn't a valid value in the j5 component spec anyway.

https://github.com/j5api/j5/blob/main/j5/components/piezo.py#L145-L146

Tracking this in #820

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py:102]: e4c482968c25037946c69e16290c6a30ce2dd488

File: j5/backends/hardware/sr/v4/serial/protocol.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/protocol.py:180]: 1e90930f019bb625177ee5a3e84e1634773eea4f

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py:153]: bac537b791ef9b1abb40f44816065fda5de370a8

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py:139]: bac537b791ef9b1abb40f44816065fda5de370a8

**Comment by trickeydan** [comment]
Fixes #816

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [reply j5/backends/hardware/sr/v4/serial/power_board.py:188]: This seems like a substantial shortcoming given that sending 0 frequency buzzes was how the old firmware did gaps in tones.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [reply j5/backends/hardware/sr/v4/serial/power_board.py:102]: This commit doesn't seem to be in this PR

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (WillB97) [inline j5/backends/hardware/sr/v4/serial/power_board.py:155]: Just realised that we aren't handling float throwing value errors throughout this file.

File: j5/backends/hardware/sr/v4/serial/power_board.py
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/serial/power_board.py:199]: This will have to be addressed later down the line. We don't have time at the moment and I'm afraid I don't agree that this is substantial.

**Pull request merged**
